In [393]:
# import packages

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


# Exploratory Data Analysis

## 1 ) Transaction Data

In [394]:
# Load the data
transaction = pd.read_excel(r"/content/QVI_transaction_data.xlsx")

In [395]:
transaction

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0       43390          1            1000       1         5   
1       43599          1            1307     348        66   
2       43605          1            1343     383        61   
3       43329          2            2373     974        69   
4       43330          2            2426    1038       108   
...       ...        ...             ...     ...       ...   
264831  43533        272          272319  270088        89   
264832  43325        272          272358  270154        74   
264833  43410        272          272379  270187        51   
264834  43461        272          272379  270188        42   
264835  43365        272          272380  270189        74   

                                       PROD_NAME  PROD_QTY  TOT_SALES  
0         Natural Chip        Compny SeaSalt175g         2        6.0  
1                       CCs Nacho Cheese    175g         3        6.3  
2         Smiths Crinkle Cut  Chips Chicken 170g         2        2.9  
3         Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0  
4       Kettle Tortilla ChpsHny&Jlpno Chili 150g         3       13.8  
...                                          ...       ...        ...  
264831   Kettle Sweet Chilli And Sour Cream 175g         2       10.8  
264832             Tostitos Splash Of  Lime 175g         1        4.4  
264833                  Doritos Mexicana    170g         2        8.8  
264834   Doritos Corn Chip Mexican Jalapeno 150g         2        7.8  
264835             Tostitos Splash Of  Lime 175g         2        8.8  

[264836 rows x 8 columns]

In [396]:
transaction.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 264836 entries, 0 to 264835
Data columns (total 8 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   DATE            264836 non-null  int64  
 1   STORE_NBR       264836 non-null  int64  
 2   LYLTY_CARD_NBR  264836 non-null  int64  
 3   TXN_ID          264836 non-null  int64  
 4   PROD_NBR        264836 non-null  int64  
 5   PROD_NAME       264836 non-null  object 
 6   PROD_QTY        264836 non-null  int64  
 7   TOT_SALES       264836 non-null  float64
dtypes: float64(1), int64(6), object(1)
memory usage: 16.2+ MB


In [397]:
# Check TXN_ID duplicate ID
transaction.TXN_ID.duplicated().sum()

1709

##### Missing data

In [398]:
transaction.isnull().sum()

DATE              0
STORE_NBR         0
LYLTY_CARD_NBR    0
TXN_ID            0
PROD_NBR          0
PROD_NAME         0
PROD_QTY          0
TOT_SALES         0
dtype: int64

- Zero null value in Our data provided

#### PROD_NAME
### PROD_NAME
- PROD_NAME is kind of long phrases with nonsense line, in this part I ahould be better to clean it and reduce the size of each item to one or two words with meaningful items
- As we are only interested in words that will tell us if the product is chips or not, let’s remove all words with
digits and special characters such as ‘&’ from our set of product words.



In [417]:

# remove the letter 'g' at the end of each row and all non-alphabetic characters
transaction['PROD_NAME'] = transaction['PROD_NAME'].str.replace(r'[^a-zA-Z]+g?G?\s*$', '', regex = True).str.replace("S/","").str.replace("&","").str.replace(r"\d+","")

<ipython-input-417-91747966fe3c>:2: FutureWarning: The default value of regex will change from True to False in a future version.
  transaction['PROD_NAME'] = transaction['PROD_NAME'].str.replace(r'[^a-zA-Z]+g?G?\s*$', '', regex = True).str.replace("S/","").str.replace("&","").str.replace(r"\d+","")


In [418]:
# Checking 10 rows
transaction.head(10)

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0  43390          1            1000       1         5   
1  43599          1            1307     348        66   
2  43605          1            1343     383        61   
3  43329          2            2373     974        69   
4  43330          2            2426    1038       108   
5  43604          4            4074    2982        57   
6  43601          4            4149    3333        16   
7  43601          4            4196    3539        24   
8  43332          5            5026    4525        42   
9  43330          7            7150    6900        52   

                             PROD_NAME  PROD_QTY  TOT_SALES  
0   Natural Chip        Compny SeaSalt         2        6.0  
1                     CCs Nacho Cheese         3        6.3  
2    Smiths Crinkle Cut  Chips Chicken         2        2.9  
3       Smiths Chip Thinly  CreamOnion         5       15.0  
4   Kettle Tortilla ChpsHnyJlpno Chili         3       13.8  
5  Old El Paso Salsa   Dip Tomato Mild         1        5.1  
6   Smiths Crinkle Chips Salt  Vinegar         1        5.7  
7     Grain Waves         Sweet Chilli         1        3.6  
8   Doritos Corn Chip Mexican Jalapeno         1        3.9  
9      Grain Waves Sour    CreamChives         2        7.2

 - Date column became more readable an more useful with Date format which can be used to do time series forecasting and analysis throughout the time

In [ ]:
# Covert DATE from digit to a date format
transaction['DATE'] = pd.to_datetime(pd.to_numeric(transaction['DATE'],errors='coerce'),unit = "D", origin = "1899-12-30")
transaction["DATE"]

#### STORE_NBR

In [ ]:
# The best 10 stores listed
print(transaction["STORE_NBR"].value_counts()[:10])
# the best 10 Store Number graph
transaction["STORE_NBR"].value_counts()[:10].plot.bar(title= "THE BEST 10 STORES")

- It can be seen that the *Store* with number **226** is the most frequent store with **2022** times followed by the **Store 88** with **1873** times

#### PRODUCT_NBR

In [ ]:
len(transaction["PROD_NBR"].unique())

114 Is the number of the unique Products with no duplication

- There were *114* unique purchased product

In [ ]:
# The best 10 Products and most purchased
print(transaction["PROD_NBR"].value_counts()[:5])
transaction["PROD_NBR"].value_counts()[:5].plot.bar(color = 'r')

The most purchased products are, at first **102 108 33 112 75 **  with more than 3000 times

#### PROD_NBR	PROD_NAME

## Let's get the right name of each product

In [ ]:
prod = pd.Series(transaction.PROD_NBR.unique())  #.nunique()

In [ ]:
name=pd.Series(transaction.PROD_NAME.unique())   #.nunique()

In [ ]:
ls = []

for x in range(len(prod)):
  y = name[x]
  z = prod[x]
  lst = [y,z]
  ls.append(lst)




In [ ]:
# list of 10 of each product with the number representative
ls[:10]

In [ ]:
len(ls)

All the work above was done to ensure that each product ahas a unique PROD_NBR

#### TOT_SALE Distribution

In [ ]:
transaction.describe()

#### Detecting outliers

- 1_ PROD_QTY

PRO_QTY is my choice to detect outliers and as we can notice in the table above that the maximum in one transaction is 200 in the other hand the 1st,2nd, 3rd Quartiles are  much lower than 200, which means that 200 is a sign of outlier, lets use some functions to confirm

In [ ]:
#Scatter  to visuailze how the data distributed
plt.scatter(transaction["PROD_QTY"],transaction["TOT_SALES"])


- It is abvious in the illustration above that we have transactions with  PROD_QTY = 200
- in the next part We make visualization with and without outliers concedering the transaction of 200 as outlier and make comparison of how outliers can effect the data
- At first. Let's extract the costumer and how many times the purchase happened   

In [ ]:
outliers_row  = transaction.loc[transaction['PROD_QTY'] == 200]
outliers_row

The extraction above shows that the customer bought Chips in two different days, the customer hold the loyality card (226000)

In [ ]:
# Let's see if the customer has had any other transactions with 226000 Loyality_card_number

outliers_row  = transaction.loc[transaction['LYLTY_CARD_NBR'] == 226000]
outliers_row

It looks like this customer has only had the two transactions over the year and is not an ordinary retail
customer. The customer might be buying chips for commercial purposes instead. We’ll remove this loyalty
card number from further analysis.


In the graph above we can see how outliers effect our Data and the analysis, which means We should be aware about Outliers in each case, and how the data normally distributed which for sure can be more helpful for our analysis instead of data with outliers

### Working on removing outliers

> Because of removing outliers is very sensitive in Data analysis,Outliers can be detected using visualization, implementing mathematical formulas on the dataset, or using the statistical approach.l ike z_score, IQR Q! and Q3,but in our case We want just to remove the customer with 200 and Loyality card number of 226000, this approach is considered the best to get good results and insights.





In [ ]:
# Box Plot to showcase outliers
import seaborn as sns
sns.boxplot(transaction['PROD_QTY'])



In [ ]:
# Remove Outliers
# In this case, We use loc command instead of complicated code

transaction  = transaction.loc[transaction['LYLTY_CARD_NBR'] != 226000]
transaction

In [ ]:
# The code below shows how outliers removed from the DataFrame
outliers_check = transaction.loc[transaction['LYLTY_CARD_NBR'] == 226000]
outliers_check

In [ ]:
# DataFrame summary after removing outliers

transaction.describe()

- Depending on the summary it looks better now to move to the next step where we can work with DATES to get more insights

In [ ]:
# How much unique days We have in our dataset
transaction["DATE"].nunique()

- The dataset is approximately one year


In [ ]:
# lets check missing values again after outliers removal and DATES types conversion
transaction.isnull().sum()

In [ ]:
transaction["DATE"] = pd.to_datetime(transaction["DATE"], unit='D')
transaction.sort_values(by = "DATE", inplace = True)
transaction.head()

In [ ]:
# Using a inbuilt style to change
# the look and feel of the plot
plt.style.use("fivethirtyeight")

# setting figure size to 12, 10
plt.figure(figsize=(10, 8))

transaction_per_time = transaction.groupby(transaction.DATE)['TOT_SALES'].sum()
# Labelling the axes and setting
# a title
plt.xlabel("Date")
plt.ylabel("PROD_QTY")
plt.plot(transaction_per_time, linewidth = .6)

We can see that there is an increase in purchases in December (The End of the year) and a break in late December. And decrease in purchases in the last days of August 2018 and May 2019

In [ ]:
transaction_T_P = trnasaction[""]
transaction.groupby(transaction.DATE).sum().plot(subplots=True, figsize=(12, 15))


#### ROD_NBR	 & PROD_QTY

In [ ]:
group_PROD_NBR = transaction[['PROD_NBR','PROD_QTY']]
grouped_PROD = group_PROD_NBR.groupby(['PROD_NBR']).sum()
grouped_PROD.sort_values(['PROD_QTY'],ascending=False)[:10]

In [ ]:
group_PROD_NBR = transaction[['PROD_NBR','PROD_QTY']]
grouped_PROD = group_PROD_NBR.groupby(['PROD_NBR']).sum()
grouped_PROD.sort_values(['PROD_QTY'],ascending=False)[:10].plot(kind = "pie", subplots= True,figsize = (6,6),legend = False
                                                                ,title = "Illustration of the best 10th Saled Production and their QTY  ")

#### ROD_NBR, PROD_QTY and TOT_SALES

In [ ]:
group_PROD_NBR_SALES = transaction[['PROD_NBR','PROD_QTY','TOT_SALES']]
grouped_PROD_SALES = group_PROD_NBR_SALES.groupby(['PROD_NBR']).sum()
grouped_PROD_SALES.sort_values(['TOT_SALES'],ascending=False)[:10]

- It can be seen that the data contains outliers and the scatter plot above illustrates that the PROD_QTY with 200 is an outlier value, but style rough to decide if we can drop it or not

####################


## 2) Purchase_behaviour -----------------------------------

This part is specified to analyse the second dataset of **purchased_behaviour**

In [ ]:
purchase_behaviour = pd.read_csv(r"/content/QVI_purchase_behaviour.csv")
purchase_behaviour.head(20)

In [ ]:
purchase_behaviour.info()

In [ ]:
# Check duplicate card numbers
purchase_behaviour.LYLTY_CARD_NBR.duplicated().sum()

In [ ]:
# Missing Values
purchase_behaviour.isnull().sum()

- ZERO missing values

In [ ]:
print(purchase_behaviour['LIFESTAGE'].value_counts())
purchase_behaviour['LIFESTAGE'].value_counts().plot.pie(figsize=(7,7),title = "Count of all lifestages exist",legend = False,
                                                       ylabel = "");

- obvious that LIFESTAGE contains 7 classes

In [ ]:
# PREMIUM_CUSTOMER
print(purchase_behaviour['PREMIUM_CUSTOMER'].value_counts())
purchase_behaviour['PREMIUM_CUSTOMER'].value_counts().plot.bar(figsize=(7,7),title = "Count of all PREMIUM CUSTOMERS exist",legend = False,
                                                       ylabel = "",rot = 0);

Depending on the illustration above of the bar graph Mainstream has the high amount with 29245 purhasing time fllowed by Budget and Premium with 24470  and 18922 respectively

In [ ]:
(purchase_behaviour[['LIFESTAGE','PREMIUM_CUSTOMER']].groupby('LIFESTAGE').count()).plot.barh(legend = False)

In [ ]:
purchase_behaviour[['LIFESTAGE','PREMIUM_CUSTOMER']].groupby(['PREMIUM_CUSTOMER','LIFESTAGE'])['PREMIUM_CUSTOMER'].count()